Move embedded reply in posts to a reply object

teceler opened the pull request:
Fixes #421 

~~Depends on #667~~

Depends on #1351 and #1331

teceler committed:
Create migration

teceler committed:
Render first reply in previous location & hide it from normal reply list

teceler committed:
Clean up various post-vs-replies special-casing

teceler committed:
Initialize writable

teceler committed:
Add written_params

teceler committed:
Remove Post from Writable Concern

And add `editable_by?` and `deletable_by?` to post model

teceler committed:
Use fields_for to create nested form for writable content

teceler committed:
Correct editor and preview loading of attributes

teceler committed:
Update factorybot post creation

teceler committed:
Make various tests use replies

teceler committed:
Make various tests using post content use post fields

teceler committed:
Fix controller calls on posts_controller_spec

teceler committed:
Remove outdated tests

teceler committed:
Fix reply.ordered tests

teceler committed:
Fix replies#search tests

teceler committed:
Fix paginators to ignore first reply

teceler committed:
Make test find the right reply

teceler committed:
Stop search checking post content

teceler committed:
Tell build template groups to look at the reply object on post preview

teceler committed:
Handle firstposts correctly in replies/_single.haml

teceler committed:
Expect thread characters to exist in reply previews

teceler committed:
Make reply controller reply update preview test take into account thread characters

Throne3d committed:
Fix some FlatPost tests

teceler committed:
Adjust api tests (review this! changes api functionality)

teceler committed:
Fix new tests

teceler committed:
Update timestamp tests

teceler committed:
Exclude first reply when paginating

teceler committed:
Fix marking unread regression

teceler committed:
Minor cleanup

Marri committed:
Use Rails associations rather than a method to support Rails magic

teceler committed:
Update selectors in feature specs

Marri committed:
Refactor @written variable away

Marri commented on app/controllers/posts_controller.rb:
```diff
@@ -91,10 +91,12 @@ def unhide
     @post.board_id = params[:board_id]
     @post.section_id = params[:section_id]
-    @post.icon_id = (current_user.active_character ? current_user.active_character.default_icon.try(:id) : current_user.avatar_id)
+    @post.written = @post.replies.new(user: current_user, post_id: @post.id, reply_order: 0)
```
Can we not use `@post.written.build`, did we try that already? Honestly build vs new doesn't matter a ton, but also I'm pretty sure reply_order should auto fill, and if it doesn't that seems like a Thing to Fix?

teceler replied:
it looks like it doesn't set the reply_order until it/the post actually saves, which presumably causes problems in preview?

Marri committed:
Don't assign associations to posts when updating if the preview parameter is present

Marri commented on app/controllers/icons_controller.rb:
```diff
@@ -45,8 +45,7 @@ def show
       arel = Post.arel_table
-      post_ids = Reply.where(icon_id: @icon.id).select(:post_id).distinct.pluck(:post_id)
-      where_calc = arel[:icon_id].eq(@icon.id).or(arel[:id].in(post_ids))
+      where_calc = arel[:id].in(Reply.where(icon_id: @icon.id).select(:post_id).distinct.pluck(:post_id))
```
We can leave this if you want but it's now really just `post_ids = Reply.where(icon_id: @icon.id).select(:post_id).distinct.pluck(:post_id)` and then `Post.where(id: post_ids)`, we don't need any of this arel stuff any more.

Marri commented on app/controllers/icons_controller.rb:
```diff
@@ -87,8 +86,7 @@ def replace
 
-    post_ids = Reply.where(icon_id: @icon.id).select(:post_id).distinct.pluck(:post_id)
-    all_posts = Post.where(icon_id: @icon.id) + Post.where(id: post_ids)
+    all_posts = Post.where(id: Reply.where(icon_id: @icon.id).select(:post_id).distinct.pluck(:post_id))
```
instead of `all_posts = [thing]`, remove following line and use `@posts = [thing]`, we already did distinct to dedupe for us (and Rails will dedupe duplicate ids for you anyway, `Post.where(id: [1,1,2])` returns two objects)

Marri commented on app/controllers/posts_controller.rb:
```diff
@@ -106,24 +108,28 @@ def new
+    rescue ActiveRecord::RecordInvalid
       flash.now[:error] = {}
       flash.now[:error][:array] = @post.errors.full_messages
       flash.now[:error][:message] = "Your post could not be saved because of the following problems:"
```
Did we check this includes reply validation errors?

teceler replied:
We don't currently have a way to get an invalid reply in this context, since the user will always be the same as the post user and that's all we currently validate! I'll add some code to make sure we get reply errors if any exist though.

Marri commented on app/controllers/posts_controller.rb:
```diff
@@ -106,24 +108,28 @@ def new
-    unless @post.save
+    begin
+      @post.save!
+      flash[:success] = "You have successfully posted."
```
I'd leave this at the end; minimizing stuff in a begin/rescue is good practice unless we need to.

Marri commented on app/controllers/posts_controller.rb:
```diff
@@ -149,15 +155,17 @@ def update
     labels = process_tags(Label, :post, :label_ids)
 
-    if current_user.id != @post.user_id && @post.audit_comment.blank? && !@post.author_ids.include?(current_user.id)
+    if current_user.id != @post.user_id && @post.written.audit_comment.blank? && @post.audit_comment.blank? && !@post.author_ids.include?(current_user.id)
```
Why do we have two of these...? Do we give two boxes for edit notes when you update a post?

teceler replied:
...I don't know. ...I might have just assumed we'd have one tbqh

Marri commented on app/controllers/posts_controller.rb:
```diff
@@ -234,7 +243,7 @@ def search
       arel = Post.arel_table
       post_ids = Reply.where(character_id: params[:character_id]).select(:post_id).distinct.pluck(:post_id)
-      where = arel[:character_id].eq(params[:character_id]).or(arel[:id].in(post_ids))
+      where = arel[:id].in(post_ids)
```
No longer needs to be arel

Marri commented on app/controllers/templates_controller.rb:
```diff
@@ -29,7 +29,7 @@ def show
     post_ids = Reply.where(character_id: character_ids).select(:post_id).distinct.pluck(:post_id)
     arel = Post.arel_table
-    where = arel[:character_id].in(character_ids).or(arel[:id].in(post_ids))
+    where = arel[:id].in(post_ids)
```
No longer needs arel

Marri commented on app/controllers/users_controller.rb:
```diff
@@ -114,10 +114,9 @@ def output
     @replies = Reply.where(user: current_user).where('created_at between ? AND ?', daystart, dayend).pluck(:content)
 
-    @total = @posts + @replies
+    @total = @replies
```
This can probably be cleaned up at least a little now that we don't need to do array addition

Marri commented on app/controllers/writable_controller.rb:
```diff
@@ -66,8 +67,11 @@ def show_post(cur_page=nil)
       self.page = cur_page = cur_page.to_i
     end
 
+    read_at = @post.read_time_for(@replies) if logged_in?
```
Why did this get moved? I was reasonably sure this needed to happen after paginate?

teceler replied:
...looking at the intervening code it might be so it correctly includes the written in the replies to check for read status. otoh you probably... want pagination there, yes. hm.

Marri commented on app/views/icons/show.haml:
```diff
@@ -96,7 +95,7 @@
       %tr
         %th.sub.width-150 Times Used
-        %td.even=(posts_using.count + replies_using.count).to_s
+        %td.even=(replies_using.count).to_s
```
Probably does not require ()s anymore

Marri commented on app/views/icons/show.haml:
```diff
@@ -96,7 +95,7 @@
       %tr
         %th.sub.width-150 Posts In
-        %td.even= (posts_using.pluck(:id) + replies_using.pluck('distinct post_id')).uniq.count
+        %td.even= (replies_using.pluck('distinct post_id')).uniq.count
```
`replies_using.select(:post_id).distinct.count`

Marri commented on db/migrate/20180209163253_split_post_text_into_replies.rb:
```diff
@@ -0,0 +1,45 @@
+      else
+        raise "Post user does not match initial reply's user"
+      end
+      # mass-shift the reply order here to avoid it doing them one at a time
```
ping

Marri commented on spec/models/template_spec.rb:
```diff
@@ -53,7 +53,7 @@
       character2 = create(:character, template: template, template_name: "nickname")
       character3 = create(:character, template: template, screenname: "screen_name")
-      expect(template.plucked_characters).to eq([[character1.id, character1.name], [character2.id, "#{character2.name} | #{character2.template_name}"], [character3.id, "#{character3.name} | #{character3.screenname}"]])
+      expect(template.plucked_characters).to match_array([[character1.id, character1.name], [character2.id, "#{character2.name} | #{character2.template_name}"], [character3.id, "#{character3.name} | #{character3.screenname}"]])
```
Is plucked_characters not a consistent order?

teceler replied:
I seem to have lost this change in the rebase and don't have any new failures there, so I... guess... it is? idk why I thought it wasn't.

Marri commented on spec/factories.rb:
```diff
@@ -41,19 +41,29 @@
     board
-    description { "" }
-    content { "test content" }
+    description ""
```
Aren't all values supposed to be in {}s? I thought that was a change we got bothered into by something?

Marri commented on spec/controllers/replies_controller_spec.rb:
```diff
@@ -622,13 +629,13 @@
       reply_post.mark_read(reply_post.user)
       put :update, params: { id: reply.id, reply: {content: 'new content'} }
-      expect(reply.reload.reply_order).to eq(1)
+      expect(reply  .reload.reply_order).to eq(order)
```
Why space, why new variable?

Marri commented on spec/controllers/replies_controller_spec.rb:
```diff
@@ -334,11 +340,12 @@
     it "sets reply_order correctly on the first reply" do
       reply_post = create(:post)
       login_as(reply_post.user)
+      reply_post.mark_read(reply_post.user)
```
Why is this necessary?

teceler replied:
Because otherwise it goes 'OH NO YOU HAVE AN UNREAD REPLY' when you try to add one >.>

teceler replied:
We didn't need to do that before because the code to check for unread replies doesn't trip on the _post_. I guess I could change it to ignore the first reply too.

teceler commented on app/controllers/characters_controller.rb:
```diff
@@ -185,7 +185,7 @@ def replace
     @alt = @alts.first
 
-    all_posts = Post.where(character_id: @character.id) + Post.where(id: Reply.where(character_id: @character.id).select(:character_id).distinct.pluck(:post_id))
+    all_posts = Post.where(id: Reply.where(character_id: @character.id).select(:character_id).distinct.pluck(:post_id))
```
...I think I have another PR to fix that actually xD

Marri commented:
down migration needs search indices added back in

teceler committed:
Test post#create rescues invalid first reply

teceler committed:
Add any reply errors to flash[:error] on post#create and post#update

teceler committed:
Cut now-unused post.last_updated

Throne3d commented on app/controllers/characters_controller.rb:
```diff
@@ -378,8 +373,7 @@ def og_data
     desc << generate_short(@character.description) if @character.description.present?
-    reply_posts = Reply.where(character_id: @character.id).select(:post_id).distinct.pluck(:post_id)
-    posts_count = Post.where(character_id: @character.id).or(Post.where(id: reply_posts)).uniq.count
+    posts_count = Post.where(id: Reply.where(character_id: @character.id).select(:post_id).distinct.pluck(:post_id)).count
```
```rb
Reply.where(character_id: @character.id).select(:post_id).distinct.pluck(:post_id)
```
will by itself provide a list of distinct post IDs with the character, so we can just put a count on that (and `.distinct.count` might work also, I dunno)

teceler replied:
oh that's clever. yeah that makes sense

Throne3d commented on app/services/post_scraper.rb:
```diff
@@ -152,11 +152,12 @@ def import_post_from_doc(doc)
     content = doc.at_css('.entry-content').inner_html
 
     @post = Post.new
+    @written = @post.replies.new
```
Do we need `@post.written = @written = @post.replies.new` here? I think we do that in the controller.

teceler replied:
I have no idea

teceler replied:
we do apparently

Throne3d commented on app/services/post_scraper.rb:
```diff
@@ -167,12 +168,14 @@ def import_post_from_doc(doc)
     end
 
     set_from_username(@post, username)
```
`set_from_username` tries to set a character, which will fail on a `@post`, right? Should we `set_from_username(@written, username)` and then `@post.user = @written.user`?

Throne3d replied:
Oh. You check if it's a Post later. I guess.

teceler replied:
Huh, I wonder why the tests didn't catch that

Throne3d commented on app/views/posts/_write.haml:
```diff
@@ -11,18 +11,22 @@
+          = w.hidden_field :icon_id, id: 'reply_icon_id'
+          = w.hidden_field :character_alias_id, id: 'reply_character_alias_id'
+          = submit_tag (post.new_record? ? 'Post' : 'Save'), class: 'button', id: 'submit_button', data: { disable_with: 'Saving...' }
+          = submit_tag "Preview", class: 'button', id: 'preview_button', name: 'button_preview', data: { disable_with: 'Previewing...' }
```
Looks like we have duplicate buttons between here and the next few lines?

Throne3d replied:
Added 57464cf3938162e8b7d20ab5eb9c46103829ba16, which should hopefully fix this!

Throne3d commented on app/views/posts/_write.haml:
```diff
@@ -79,4 +83,4 @@
     .subber.centered= yield :buttons
   - else
-    = render 'editor', item: post
+    = render 'editor', item: post.replies.first
```
`post.written`*, for consistency? Plus this might not inherently order `post.replies` so we may have issues with arbitrary database ordering.

teceler replied:
I think that failed on preview somehow

Throne3d replied:
Can we solve that in the controller, by forcibly setting `post.written` if it's not already set there? That way it avoids some of the other pitfalls that this might here.

Throne3d commented:
I'm not confident that post history is handled correctly:

https://github.com/Marri/glowfic/blob/c78d4df59d88355c315f22413344a8f6b7634880/app/views/writable/_history.haml#L65

Looks like we're gonna be trying to render a `@post` using `replies/single`?

Throne3d commented on app/views/replies/_single.haml:
```diff
@@ -1,7 +1,8 @@
-  - write = 'post-post'.freeze
-- elsif reply.is_a?(Reply)
-  - write = 'post-reply'.freeze
+- if reply.is_a?(Reply)
```
This should _only_ take a reply – especially given that, on line 16, we have an unguarded reference to `reply.icon_id`, which won't exist on a post. Right? (See also <https://github.com/Marri/glowfic/pull/705#issuecomment-513939874>)

Throne3d commented on app/views/replies/_single.haml:
```diff
@@ -32,13 +33,20 @@
+        - if reply.order == 0 && reply.post.metadata_editable_by?(current_user)
+          = link_to path_for(reply.post, 'edit_%s') do
+            = image_tag "icons/pencil.png".freeze, title: 'Edit'.freeze, alt: 'Edit'.freeze
+        - elsif reply.editable_by?(current_user)
```
I don't love the duplication of this "Edit" button, can we re-combine the condition? Either the reply is editable by the user, or the "reply" is the `post.written` and the post is editable by the user.

Or, I guess, we could segue this into fixing #218 sometime, using a different pencil? Maybe leave a comment in the codebase about why it's split up?

Throne3d replied:
…Oh, I see why it's separate – different paths. That's fair. Never mind.

Throne3d commented on app/views/replies/_single.haml:
```diff
@@ -32,13 +33,20 @@
+        - elsif reply.editable_by?(current_user)
+          = link_to path_for(reply, 'edit_%s') do
+            = image_tag "icons/pencil.png".freeze, title: 'Edit'.freeze, alt: 'Edit'.freeze
+        - if reply.reply_order == 0  && reply.post.deletable_by?(current_user)
```
(Super minor: There's a double space before the `&&`.)

Throne3d commented on db/migrate/20180209163253_split_post_text_into_replies.rb:
```diff
@@ -0,0 +1,50 @@
+class SplitPostTextIntoReplies < ActiveRecord::Migration[5.1]
+  def up
+    Post.find_each do |post|
+      post.replies.ordered.each { |reply| reply.update_columns(reply_order: reply.reply_order+1) }
```
This is going to be super slow. Uhhh, since this all goes on in a transaction, put something like this outside the loop above? (I'd appreciate it being tested in practice :P)

```rb
execute <<-SQL
UPDATE replies
SET reply_order = reply_order + 1;
SQL
```

teceler replied:
I was kinda assuming someone was gonna do SQL Magic to it before it was released yeah

Throne3d commented on db/migrate/20180209163253_split_post_text_into_replies.rb:
```diff
@@ -0,0 +1,50 @@
+        updated_at: post.updated_at,
+        skip_post_update: true
+      )
+      reply.update_columns(reply_order: 0, created_at: post.created_at)
```
Might want to set `updated_at` to be same as the post's `edited_at` (or whatever the right one is), so we don't spawn a ton of history changes.

Throne3d commented on db/migrate/20180209163253_split_post_text_into_replies.rb:
```diff
@@ -0,0 +1,50 @@
+    change_table :posts do |t|
+      t.remove :character_id, :character_alias_id, :icon_id, :content
+    end
+    Post.auditing_enabled = true
```
Does this need auditing disabled?

teceler replied:
otherwise we spawn an audit for every. single. post. for those parts being deleted and moved to the reply. If you wanna re-enable it for prod that's sensible but I got fed up with it

Throne3d replied:
Huh! I wouldn't have expected audits to act here, since it doesn't have a `current_user`. For prod, I think I'd like it enabled, so we can preserve data alright, but I'm not sure.

Throne3d commented on db/migrate/20180209163253_split_post_text_into_replies.rb:
```diff
@@ -0,0 +1,50 @@
+        raise "Post user does not match initial reply's user"
+      end
+      post.replies.ordered.each { |reply| reply.update_columns(reply_order: reply.reply_order-1) }
+      reply.destroy!
```
I'd use `reply.delete` so we don't run callbacks.

I'd also execute SQL for the replies' reply_order, outside the end of this loop:

```
execute <<-SQL
UPDATE replies
SET reply_order = reply_order - 1;
SQL
```

Throne3d commented on spec/controllers/api/v1/replies_controller_spec.rb:
```diff
@@ -17,7 +17,7 @@
 
     it "succeeds with valid post", :show_in_doc do
-      post = create(:post, num_replies: 2, with_icon: true, with_character: true)
+      post = create(:post, num_replies: 1, with_icon: true, with_character: true)
```
This test suggests we're now showing `post.written` in the API, where we didn't before. Is that correct behavior?

teceler replied:
I have _no idea_

Throne3d replied:
I'm okay with this change, I think. Looks like our current docs don't show the post as being part of the replies endpoint, but it _also_ doesn't show the post content as being part of the _posts_ endpoint, and we need it somewhere.

Should let Vona know though (they maintain glowficrss.com, right?).

Throne3d commented on spec/controllers/icons_controller_spec.rb:
```diff
@@ -499,15 +499,15 @@
         icon = create(:icon, user_id: user.id)
         alts = Array.new(5) { create(:icon, user_id: user.id) }
-        post = create(:post, user_id: user.id, icon_id: icon.id)
+        post = create(:post, user_id: user.id, icon: icon)
```
We're creating a post with a particular icon? Does this need to be changed?

Throne3d replied:
Never mind, I see we've amended the factory.

Throne3d commented on spec/controllers/icons_controller_spec.rb:
```diff
@@ -499,15 +499,15 @@
         expect(gallery.icons).to match_array([other_icon])
-        create(:post, user_id: user.id, icon_id: other_icon.id) # other post
-        create(:reply, user_id: user.id, icon_id: other_icon.id) # other reply
+        create(:post, user_id: user.id, icon: other_icon) # other post
```
Same as earlier

Throne3d commented on spec/controllers/posts_controller_spec.rb:
```diff
@@ -106,12 +107,6 @@
       end
 
-      it "does not mix up subject with content" do
-        create(:post, subject: 'unrelated', content: 'contains stars')
```
Shouldn't we still have this test, just adding a reply in?

teceler replied:
I wouldn't expect that to be a thing it could be confused about, with a reply?

Throne3d replied:
Probably not! Hmmmmmmmm. Okay I'm fine with this out.

Throne3d commented on spec/controllers/posts_controller_spec.rb:
```diff
@@ -638,6 +661,10 @@
     end
 
+    it "handles invalid first reply" do
+      # TODO: Add a test for this if it becomes possible to have an invalid reply here
```
This should explicitly say `skip` in it, as currently it'll act as a passing test.

Throne3d commented on spec/controllers/posts_controller_spec.rb:
```diff
@@ -944,7 +975,7 @@
 
       it "works for unread" do
-        third_reply = post.replies.ordered.limit(3).last
+        third_reply = post.replies.ordered.limit(4).last
```
*squints* I think this is right? Looks a bit confusing, though, maybe we should just do:
```rb
replies = post.replies.ordered.to_a
third_reply = replies[3]
second_last_reply = replies[-2]
```
or something like that?

Throne3d commented on spec/controllers/replies_controller_spec.rb:
```diff
@@ -61,13 +61,33 @@
-        expect(template).to eq(char2.template)
+        expect(templates.length).to eq(3)
+        used = templates.first
+        expect(used.name).to eq("Thread characters")
```
Hrrrm, do we want this? It's probably a pain to work around, and it's probably okay as is.

Throne3d commented on spec/controllers/replies_controller_spec.rb:
```diff
@@ -1014,7 +1037,7 @@
         filtered_reply = replies.last
         get :search, params: { commit: true, post_id: filtered_reply.post_id }
-        expect(assigns(:search_results)).to match_array([filtered_reply])
+        expect(assigns(:search_results)).to match_array([filtered_reply.post.written, filtered_reply])
```
We want this, right? For searches to find the main post? This is good!

teceler replied:
This is the main user-facing problem with our current model, yeah!

Throne3d commented on spec/features/posts/flat_spec.rb:
```diff
@@ -9,7 +9,7 @@
     visit post_path(post, view: 'flat')
     expect(page).to have_selector('.post-container', count: 1)
-    expect(page).to have_selector('.post-post', count: 1)
+    expect(page).to have_selector('.post-reply', count: 1)
```
This seems like a fair change, but I want to note that it _is_ different from posts#show, and we should maybe look into renaming things appropriately.

Throne3d commented on spec/models/post_spec.rb:
```diff
@@ -17,15 +17,33 @@
+          expect(post.edited_at).to be > old_edited_at
+        end
+      end
+
```
How does `edited_at` differ from `updated_at`, now? I think it doesn't. Maybe we should look into removing it in a follow-up PR.

teceler replied:
I think edited_at is still updated when the metadata changes?

Throne3d commented on spec/models/post_spec.rb:
```diff
@@ -58,12 +76,21 @@
       }
 
-      it "should update edited_at but not tagged_at when content edited" do
+      it "should update edited_at and but not tagged_at when subject edited" do
```
and not tagged_at*

Throne3d commented on spec/models/post_spec.rb:
```diff
@@ -58,12 +76,21 @@
+      it "should not update when content edited" do
+        Timecop.freeze(time) do
+          post.written.content = 'newer content'
+          post.save!
```
What's up with this saving the post? Is this meant to be intentionally different from the above "should update tagged_at but not edited_at when content edited"?

teceler replied:
saving hte post should save the content, but it should _not_ change either of those values. I should probably check it does save the content tho

teceler replied:
apparently it doesn't

teceler replied:
making them the same now xD

teceler committed:
Stop writable/_history from trying to render a post as content

Throne3d commented on spec/models/post_spec.rb:
```diff
@@ -176,28 +203,6 @@
       expect(post.edited_at).to eq(post.created_at)
     end
-
-    it "should update correctly when characters are edited" do
```
Is this stuff duplicated in `reply_spec.rb`, then?

Throne3d replied:
& stuff below

teceler replied:
`#has_icons` was duplicated; we don't have the entire `#edited_at` thing on replies so that isn't duplicated

teceler committed:
minor test improvements

teceler committed:
Fix seeds

teceler committed:
Use post.written rather than post.replies.first on posts/_write

teceler committed:
Use Throne's ~SQL magic~

teceler committed:
Skip draft deletion check on written

teceler committed:
Fix timestamp test

teceler committed:
Fix RepliesController#restore tests

teceler committed:
Clean up build_template_groups

teceler commented on app/controllers/writable_controller.rb:
```diff
@@ -13,9 +13,7 @@ def build_template_groups(user=nil)
-      written = @reply || @post.written
-      uniq_chars_ids = @post.replies.where(user_id: user.id).where('character_id is not null').group(:character_id).pluck(:character_id)
-      uniq_chars_ids << written.character_id unless written.character_id.nil? or uniq_chars_ids.include?(written.character_id)
+      uniq_chars_ids = @post.replies.where(user: user).select(:character_id).distinct.pluck(:character_id)
```
Although, hmm, this is probably going to get weird results on post preview

teceler commented:
To do:
- [ ] work out how to handle audit comments for post and post written (<https://github.com/Marri/glowfic/pull/705#discussion_r247379727>)
- [ ] work out how to properly fix show_post (<https://github.com/Marri/glowfic/pull/705#discussion_r247380201>)
- [ ] make an after_create marking post written read for post user? would avoid needing quite so many mark_reads in tests (<https://github.com/Marri/glowfic/pull/705#discussion_r247381425>)
- [ ] figure out what the _hell_ is going on with the editor buttons and how to fix that (<https://github.com/Marri/glowfic/pull/705#discussion_r306014330>)
- [ ] Test thread characters on post#preview (https://github.com/Marri/glowfic/pull/705#discussion_r359662990)

teceler committed:
Handle audit comments by attaching them to both objects

teceler committed:
Fix json handling for post written

teceler committed:
Fix test

Throne3d committed:
Remove duplicate button definitions

teceler commented on app/controllers/api/v1/posts_controller.rb:
```diff
@@ -22,7 +22,7 @@ def index
     return unless (post = find_object(Post))
     access_denied and return unless post.visible_to?(current_user)
-    render json: post.as_json(include: [:character, :icon, :content])
+    render json: post.as_json
```
NTS: double check the post written shows up in replies now

teceler commented on app/controllers/writable_controller.rb:
```diff
@@ -13,8 +13,8 @@ def build_template_groups(user=nil)
-      uniq_chars_ids = @post.replies.where(user_id: user.id).where('character_id is not null').group(:character_id).pluck(:character_id)
-      uniq_chars_ids << @post.character_id if @post.user_id == user.id && @post.character_id.present?
+      uniq_chars_ids = @post.replies.where(user: user).select(:character_id).distinct.pluck(:character_id)
+      uniq_chars_ids << @reply.character_id if @reply.present?
```
Why were we checking user on the local reply before? was it for coauthors or mod edits?

redmvc replied:
Dunno, if it passes tests it seems fine by me.

teceler commented on app/models/character.rb:
```diff
@@ -62,8 +62,7 @@ def self.deletable_by?(user, user_id)
     reply_ids = replies.group(:post_id).pluck(:post_id)
-    post_ids = posts.select(:id).map(&:id)
-    @recent ||= Post.where(id: (post_ids + reply_ids).uniq).ordered
+    @recent ||= Post.where(id: reply_ids.uniq).ordered
```
This can probably be cleaned up more

teceler commented on db/migrate/20180209163253_split_post_text_into_replies.rb:
```diff
@@ -0,0 +1,59 @@
+        updated_at: post.updated_at,
+        skip_post_update: true
+      )
+      reply.update_columns(reply_order: 0, created_at: post.created_at, updated_at: post.edited_at)
```
This... is going to take a long time. I don't know if there's any way to improve that.

teceler committed:
Fix new test

teceler commented on spec/controllers/characters_controller_spec.rb:
```diff
@@ -234,7 +234,7 @@
       expect(assigns(:page_title)).to eq(character.name)
       expect(assigns(:posts).size).to eq(25)
-      expect(assigns(:posts)).to match_array(Post.where(character_id: character.id).ordered.limit(25))
+      expect(assigns(:posts)).to match_array(Post.where(id: Reply.where(character_id: character.id).select(:post_id).distinct.pluck(:post_id)).order('tagged_at desc').limit(25))
```
The order can probably be revered to `.ordered`

teceler commented on spec/controllers/replies_controller_spec.rb:
```diff
@@ -654,13 +674,13 @@
       reply_post.mark_read(reply_post.user)
       put :update, params: { id: reply.id, reply: {content: 'new content'} }
-      expect(reply.reload.reply_order).to eq(1)
+      expect(reply.reload.reply_order).to eq(order)
```
can probably do this with a not_to change?

teceler committed:
More directly acknowledge changes in reply specs

Throne3d commented:
A way we might fragmentize this:

1. Make sure no code relies on `reply_order` being zero-based
2. Create a `new_reply_order` column, and when a Reply is created/updated, set it equal to `reply_order + 1`
3. In batches (by post or otherwise), make sure `new_reply_order` is filled for all replies
4. In a single transaction, do the following renames: `reply_order` -> `old_reply_order`, `new_reply_order` -> `reply_order`, `old_reply_order` -> `new_reply_order`
5. Tear out code that updates `new_reply_order`
6. Delete column `new_reply_order` (actually `old_reply_order`)
7. Make any code that currently treats Posts as a special type of Reply skip any replies with reply_order 0 (temporarily, to prevent duplicate counting). Make code that currently uses just replies use replies skipping order 0 (this is more permanent)
8. Start creating these Reply objects with reply_order 0, on new post creation + backfilling
9. Once they're all here, start reading replies with reply_order 0 instead of / in addition to posts, where relevant
10. Remove old, unused data from posts

Not sure if I got everything here, but I think this is a good chunk of the steps.

teceler committed:
Migrate

teceler committed:
Create written association

teceler committed:
Create written on new posts

teceler committed:
Update written when updating posts where it exists

teceler committed:
Create writtens

teceler committed:
Maintain order
Ensure new replies are created sanely
Ensure reply order is updated sanely when a reply is deleted

teceler committed:
Set user when creating written

teceler committed:
Fix attributes check

teceler committed:
Update reply_order usage in tests

teceler committed:
Only add extra increment to reply_order if written does not exist

teceler committed:
Only add extra increment to reply_order if written does not exist

teceler committed:
Correct test expectations

teceler committed:
Correct test expectations

teceler committed:
Skip flat post regeneration when creating written

teceler committed:
Fix more tests

teceler committed:
Fix more tests

teceler committed:
Fix unread

teceler committed:
Fix timestamp tests

teceler committed:
Fix with_reply_count

teceler committed:
Get codeclimate to chill

teceler committed:
Get codeclimate to chill

teceler committed:
Refactor

teceler committed:
Refactor

teceler committed:
Fix test

teceler committed:
Clean up written creation handling

teceler committed:
Unhide writtens

teceler committed:
Regenerate migration

teceler committed:
Exclude written from manual reply count

teceler committed:
Clean up replies/_single code

teceler committed:
Restore update post jobs

teceler committed:
Keep post up to date

teceler committed:
Remove migration

teceler committed:
Revert schema

teceler committed:
Temporarily re-include Writable in Post

teceler committed:
Fix

teceler committed:
Fix writable/history

teceler committed:
Assign written values to post and written when creating/updating post

teceler committed:
Restore cache invalidation (probably a merge error)

teceler committed:
patch timestamp tests

teceler committed:
Tell post syncing methods to leave the post alone if it's already up to date

teceler committed:
Fix total_entries reply count call

teceler committed:
Modify migration script to copy relevant audits

teceler committed:
Fix script

teceler committed:
Add a guard clause

teceler committed:
Add a skip_written variable for testing purposes

teceler committed:
Add simple tests

teceler committed:
Remove autogenerated audit before copying

teceler committed:
Add more tests

teceler committed:
Clean up redundant code

teceler committed:
Add more audit columns parameters to script

teceler committed:
Set more flags on written create/update

teceler committed:
Add a todo note about skip_written

teceler committed:
Use before/after hooks to enable/disable auditing

teceler committed:
Fix

teceler committed:
Clean up

teceler committed:
Correct test

teceler committed:
Fix update_written

teceler committed:
Fix create_written audit handling

teceler committed:
Don't skip set_last_reply on written creation

teceler committed:
Update reply count expectations in new tests

teceler committed:
Fix timestamp test

teceler committed:
Clean up dead code

teceler committed:
Render views on audit calculation test

teceler committed:
Add dependent: :destroy

teceler committed:
Update posts split tests

teceler committed:
Clean up rubocop complaints

teceler committed:
Make post split use old reply for post.written rather than deleting it

teceler committed:
Improve scraper written initialization

teceler committed:
Fix reply#set_previous_reply_to_last

teceler committed:
Fix post_split and specs

teceler committed:
Improve replies#search specs

teceler committed:
Apply suggestions from rubocop

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

teceler committed:
fix some merge errors

teceler committed:
fixes

teceler committed:
Add frozen_string_literal

teceler committed:
Fix word count

teceler committed:
Test fixes

redmvc commented on app/controllers/characters_controller.rb:
```suggestion

```
This is no longer necessary, right?

redmvc commented on app/controllers/icons_controller.rb:
```suggestion

```
Here likewise.

redmvc committed:
Fix reordering loop

redmvc committed:
Update migration

redmvc committed:
Set up and down and fix rubocop warning

redmvc committed:
Update schema version

redmvc committed:
Add missing where.not reply orders

redmvc committed:
Shorten line

redmvc committed:
Silence rubocop

redmvc committed:
Remove incorrect order condition

redmvc commented on app/controllers/posts_controller.rb:
```diff
@@ -545,14 +560,9 @@ def permitted_params(include_associations=true)
       :authors_locked,
       :audit_comment,
       :private_note,
-      :editor_mode,
```
If we decided that we're keeping all of the reply params in the post for the moment, why are they removed from permitted_params?

teceler replied:
If you're submitting it to the post at this point something has gone badly wrong so it's better for tests to fail?

redmvc commented on app/controllers/users_controller.rb:
```diff
@@ -178,19 +178,12 @@ def output
+      .where('created_at between ? AND ?', daystart, dayend)
+      .order(post_id: :asc).ordered
+      .pluck(:content)
 
```
Why did we get rid of @posts?

teceler replied:
They'll be duplicates of reply content

redmvc commented on app/controllers/writable_controller.rb:
```diff
@@ -58,12 +58,12 @@ def show_post(cur_page=nil)
       end
     elsif cur_page == 'last'
-      self.page = cur_page = @post.replies.paginate(per_page: per, page: 1).total_pages
+      self.page = cur_page = @post.replies.where.not(reply_order: 0).paginate(per_page: per, page: 1).total_pages
```
This is unnecessary, right? @post.replies already filters this.

redmvc replied:
Or, did that get undone from the .written code?

teceler replied:
Yeah. I'm not sure why this one needs it or whether it *should* but it did get undone from the written code

redmvc replied:
This one needs it bc it's paginating, it needs to ignore the top post

redmvc commented on app/models/post.rb:
```diff
@@ -53,14 +57,14 @@ class Post < ApplicationRecord
     :search,
     against: %i(
       subject
-      content
```
If we're not deleting content from posts, should this remain?

teceler replied:
We don't want to search posts at this point for duplication reasons, right?

redmvc replied:
wdym duplication reasons? Currently posts aren't searched at all, they simply do not show up in search

redmvc committed:
Update flat posts for character alias

teceler commented on app/views/posts/flat.haml:
```diff
@@ -22,36 +22,4 @@
-                = precede ' | Updated ' do
-                  %span.post-updated=pretty_time(@post.last_updated, format: ApplicationHelper::TIME_FORMAT)
-      .flat-post-replies= @post.flat_post.content.try(:html_safe)
+      = @post.flat_post.content.try(:html_safe)
```
I thiiink we escape this at earlier stages brakeman but thanks?

redmvc committed:
Restore missing columns from removed migration

redmvc committed:
Fix reply buttons in single

redmvc committed:
Fix orderable

redmvc committed:
Fix audits

redmvc committed:
Ensure post_id exists in tags

redmvc committed:
Fix indentation

redmvc committed:
Fix tagged_at

redmvc committed:
Fix history

redmvc committed:
Fix specs

redmvc committed:
Add missing post.written

redmvc committed:
Rubocop fix

redmvc committed:
Brakeman ignore fix

redmvc committed:
Fix spec

redmvc committed:
Fix reply count

redmvc commented on spec/controllers/api/v1/posts_controller_spec.rb:
```diff
@@ -74,12 +74,9 @@
       expect(response).to have_http_status(200)
       expect(response.parsed_body['id']).to eq(post.id)
-      expect(response.parsed_body['num_replies']).to eq(0)
+      expect(response.parsed_body['num_replies']).to eq(1)
```
Why should this be 1? Aren't we excluding `written` from the reply count everywhere else in the code?

teceler replied:
...i thought we stopped excluding it here, but i might have updated some tests incorrectly

redmvc replied:
Huh actually you're right. It's only in application_controller's calculate_view_status that we're still excluding it, plus places that paginate, everywhere else we did stop excluding it. So should calculate_view_status instead not exclude it?

redmvc replied:
Yeah I removed that exclusion from there, everything's ok now.

redmvc commented on spec/models/daily_report_spec.rb:
```diff
@@ -71,7 +71,7 @@
       end
       report = DailyReport.new(now)
-      expect(report.posts.first.reply_count).to eq(5)
+      expect(report.posts.first.reply_count).to eq(6)
```
Likewise here

redmvc commented on spec/models/post_spec.rb:
```diff
@@ -1131,7 +1060,7 @@
           tagged_at: post.tagged_at,
           status: :active,
-          num_replies: 0,
+          num_replies: 1,
```
And here.

redmvc committed:
Undo reply count change

redmvc committed:
Remove calculate_view_status unread counts filter

redmvc commented on app/models/post.rb:
```diff
@@ -89,7 +93,7 @@ class Post < ApplicationRecord
 
   scope :with_reply_count, -> {
-    select('(SELECT COUNT(*) FROM replies WHERE replies.post_id = posts.id) AS reply_count')
+    select('(SELECT COUNT(*) FROM replies WHERE replies.post_id = posts.id AND replies.reply_order > 0) AS reply_count')
```
@teceler seems like here we're filtering the written out of reply_count

redmvc committed:
Standardise reply counts

teceler committed:
Merge branch 'main' into fix/posts-to-replies
